test step secret inject

mjudeikis opened the pull request:
This will allow to inject secrets into test step pods. So we can test cloud interaction from test pods without the need to run complex template.

This is needed for openshift on azure work, so we could test cloud api's behaviour in test steps and keep all the logic inside openshift-azure repo, instead of release repo. 

/cc @jim-minter @stevekuznetsov

stevekuznetsov commented:
/cc @bbguimaraes @petr-muller

droslean commented on pkg/defaults/defaults.go:
```diff
@@ -144,7 +145,7 @@ func FromConfig(
 
 		} else if rawStep.TestStepConfiguration != nil {
-			step = steps.TestStep(*rawStep.TestStepConfiguration, config.Resources, podClient, artifactDir, jobSpec)
+			step = steps.TestStep(*rawStep.TestStepConfiguration, config.Resources, podClient, artifactDir, jobSpec, testSecret)
```
It also needs to be validated. see [here](https://github.com/openshift/ci-operator/blob/master/pkg/api/config.go#L90-L112)

mjudeikis replied:
So you want this to be configurable instead of flag, or both?

droslean replied:
I would expect it to be defined in [here](https://github.com/openshift/ci-operator/blob/master/pkg/defaults/defaults.go#L409)

droslean replied:
Of course, if we do this like this we won't have to make any updates in prowgen, since ci-operator will read the secret directly from its config. 

Let us decide then. @petr-muller @bbguimaraes @stevekuznetsov  Any thoughts on this one?

mjudeikis replied:
Which make sense, as now you can define secret only for steps you need, and not for all of them. Limits attach surface.

mjudeikis commented:
> but we have to inject the secret for each test from the ci-operator's configuration. 
See https://github.com/openshift/release/pull/2756/files#diff-dae6e8fc1689c5695304cb8a462cfb63R22
This is optional, but yes.

> Prowgen will have to be updated as well.
Please point to the location where updates are needed

> However, there are some security issues that we have to consider. By adding this feature, it will mean that any team can use anybody's secret and run whatever they want.

There is already ways to do this. Just spin a template and rsh to the pods and get what you need. There always be loopholes in the system. Long story short - we need this feature otherwise we will end up with 5+ more templates. And it would be easier just to move our jobs somewhere else :/

mjudeikis commented on pkg/steps/pod.go:
```diff
@@ -185,16 +151,91 @@ func TestStep(config api.TestStepConfiguration, resources api.ResourceConfigurat
+			{
+				Name:      "secret",
+				ReadOnly:  true,
+				MountPath: "/usr/secrets",
```
I like delimiters idea better than a new flag. I don't expect this to be widely used either way.

petr-muller replied:
Personally, I don't mind about these hardcodings, they make usage simple, we do not need to expose every single element to be configurable.

I would like to see the `"secret"` name to be extracted to a constant (because it needs to match in `podSecret` and `podVolume` and it might not be clear to someone who reads this a month from now) and _maaaaybe_ call the mountpoint something less less generic (`ci-operator-secrets`? `configured-secrets`?) to reduce conflict possibility. But that's probably just bikeshedding.

droslean commented:
> Please point to the location where updates are needed

[Here](https://github.com/openshift/ci-operator-prowgen/blob/master/cmd/ci-operator-prowgen/main.go#L106) prowgen fills up the command arguments in the podspec. 

An example ci-operator config should look like:
```yaml
tests:
- as: unit
  commands: make unit
  secret: mysecret
  container:
    from: src
```
Then prowgen will get the `secret` field value and appends it in the arguments.

@petr-muller @stevekuznetsov @bbguimaraes  Any thoughts?

petr-muller commented:
> @petr-muller @stevekuznetsov @bbguimaraes Any thoughts?

Haven't looked yet, looking now

petr-muller commented on pkg/steps/pod_test.go:
```diff
@@ -182,3 +183,101 @@ func TestPodStepExecution(t *testing.T) {
+			expectedPod: func(expectedPodTemplate *v1.Pod) {},
+		},
+	}
+	for _, tc := range testCases {
```
:+1: for the tests. Can we make them [subtests](https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks)?

mjudeikis replied:
are you using this pattern already somewhere?

stevekuznetsov replied:
We should be everywhere, if not, congrats, lead the charge :)

petr-muller replied:
Looks like we do :)

```
$ git grep t.Run '*_test.go' ':!vendor'
pkg/api/config_test.go:         t.Run(testCase.name, func(t *testing.T) {
pkg/api/types_test.go:          t.Run(tt.name, func(t *testing.T) {
pkg/defaults/defaults_test.go:          t.Run(testCase.name, func(t *testing.T) {
pkg/steps/artifacts_test.go:            t.Run(tt.name, func(t *testing.T) {
pkg/steps/pod_test.go:          t.Run(tc.purpose, func(t *testing.T) {
```

mjudeikis commented:
After mongling with the code I would say - lets ditch the flag and do this:
```
// gathers artifacts from that step.
type TestStepConfiguration struct {
...
	Secret *Secret `json:"secret"`
...
}

// Secret describes test secret to be mounted inside test
// container.
type Secret struct {
	Name string `json:"name"`
	Path string `json:"path"`
}
```

Simple, testable, and scoped.

mjudeikis commented:
@petr-muller @droslean check last commit I added. Removed flags, and left fully config based one. 
+split to separate secret object, as there might be more options in the future.
Tested locally too, works like a charm.

droslean commented on pkg/api/config.go:
```diff
@@ -106,6 +106,18 @@ func validateTestStepConfiguration(fieldRoot string, input []TestStepConfigurati
+			}
+			// validate path only if name is passed
+			if len(test.Secret.Path) > 0 {
+				if ok := regexp.MustCompile("^/((\\/*)\\w+)+").MatchString(test.Secret.Path); !ok {
```
is `filepath.IsAbs(test.Secret.Path)` what we need here?

droslean commented:
/test all

droslean commented on pkg/api/config_test.go:
```diff
@@ -75,7 +75,7 @@ func TestValidateTests(t *testing.T) {
 			tests: []TestStepConfiguration{
 				{
-					As: "test",
+					As:                         "test",
```
oops, you have go 1.11? The tests will fail. Please use `1.10`

stevekuznetsov replied:
eek, we need to update

mjudeikis replied:
This one is painful. Need to reconfigure my env to be dual-go-lang-version :////

droslean commented on pkg/api/types.go:
```diff
@@ -293,6 +297,13 @@ type TestStepConfiguration struct {
+// container.
+type Secret struct {
+	Name string `json:"name"`
+	Path string `json:"path"`
```
MountPath should be a better name.

droslean commented on pkg/steps/pod_test.go:
```diff
@@ -182,3 +184,135 @@ func TestPodStepExecution(t *testing.T) {
+			}
+
+			if !reflect.DeepEqual(pod, podTemplate) {
+				//if true {
```
no need for this comment.

droslean commented on pkg/steps/pod.go:
```diff
@@ -198,3 +171,82 @@ func PodStep(name string, config PodStepConfiguration, resources api.ResourceCon
+		} else {
+			secretPath = testSecretDefaultPath
+		}
+		return []coreapi.VolumeMount{
```
since we are mounting a single secret, I think we can use also the `volumeMounts.subPath`.

stevekuznetsov commented on CONFIGURATION.md:
```diff
@@ -313,6 +313,12 @@ of the images in the pipeline.
 ## `tests.container.from`
 `from` is the pipeline image tag that this test will be run on.
 
+## `tests.secret.name
```
You're missing a grave accent at the end of this line

stevekuznetsov commented on CONFIGURATION.md:
```diff
@@ -313,6 +313,12 @@ of the images in the pipeline.
 `from` is the pipeline image tag that this test will be run on.
 
+## `tests.secret.name
+`secret.name` is  secret name to be mounted inside a test container.
```
```is the name of the secret```?

stevekuznetsov commented on CONFIGURATION.md:
```diff
@@ -313,6 +313,12 @@ of the images in the pipeline.
+## `tests.secret.name
+`secret.name` is  secret name to be mounted inside a test container.
+
+## `tests.secret.path
```
Missing the trailing accent here too

stevekuznetsov commented on CONFIGURATION.md:
```diff
@@ -313,6 +313,12 @@ of the images in the pipeline.
+`secret.name` is  secret name to be mounted inside a test container.
+
+## `tests.secret.path
+`secrets.path` is an optional variable for a secret mount path. Default value `/usr/secrets`
```
`s/secrets.path/secret.path/`

stevekuznetsov replied:
`"is the path at which to mount the secret. Optional, defaults to ..."`

stevekuznetsov commented on pkg/api/config.go:
```diff
@@ -106,6 +106,18 @@ func validateTestStepConfiguration(fieldRoot string, input []TestStepConfigurati
 
+		if len(test.Secret.Name) > 0 {
+			if ok := regexp.MustCompile("^[a-zA-Z]*").MatchString(test.Secret.Name); !ok {
+				validationErrors = append(validationErrors, fmt.Errorf("%s[%d].name: '%s' secret name is not valid value, should be [a-zA-Z]", fieldRoot, num, test.Secret.Name))
```
Import the validator from upstream

mjudeikis replied:
you need to be more specific here. What validator?

droslean replied:
`k8s.io/kubernetes/pkg/apis/core/validation`

mjudeikis replied:
the price for this one was quite big. Added missing dependencies in the separate PR for easier review. Not sure if you want to do this for one regexp.

droslean replied:
hm. I would prefer the regex rather than vendoring the whole library. @petr-muller @stevekuznetsov any thoughts?

stevekuznetsov commented on pkg/api/config_test.go:
```diff
@@ -155,6 +155,64 @@ func TestValidateTests(t *testing.T) {
+					OpenshiftAnsibleClusterTestConfiguration: &OpenshiftAnsibleClusterTestConfiguration{},
+					Secret: Secret{
+						Name: "secret",
+						Path: "/path/to/secret:exec",
```
Is that not a valid filename on unix?

mjudeikis replied:
It is valid. but in some mount operations (example in docker run this is how you pass in option) it defines that mountpoint should be executable. Better safe than sorry.

stevekuznetsov commented on pkg/api/types.go:
```diff
@@ -282,6 +282,10 @@ type TestStepConfiguration struct {
 
+	// Secret is an optional secret name which
+	// will be mounted inside the test container.
+	Secret Secret `json:"secret"`
```
Should be omitempty

stevekuznetsov commented on pkg/steps/pod.go:
```diff
@@ -17,16 +17,22 @@ import (
 
 type podStep struct {
 	name        string
+	secret      string
```
Why here and also under config?

stevekuznetsov commented on pkg/steps/pod.go:
```diff
@@ -44,44 +50,9 @@ func (s *podStep) Inputs(ctx context.Context, dry bool) (api.InputDefinition, er
 func (s *podStep) Run(ctx context.Context, dry bool) error {
 	log.Printf("Executing %s %s", s.name, s.config.As)
 
-	containerResources, err := resourcesFor(s.resources.RequirementsForStep(s.config.As))
```
Moving all this makes it hard to see your diff, please don't do that

stevekuznetsov replied:
Regardless -- break it out into a separate commit or don't do it, it's not germane to the changes you are making here.

stevekuznetsov commented on pkg/steps/pod.go:
```diff
@@ -198,3 +171,82 @@ func PodStep(name string, config PodStepConfiguration, resources api.ResourceCon
+	}, nil
+}
+
+func podSecret(config PodStepConfiguration) []coreapi.Volume {
```
`s/podSecret/secretVolumeFromConfig/`?

stevekuznetsov commented on pkg/steps/pod.go:
```diff
@@ -198,3 +171,82 @@ func PodStep(name string, config PodStepConfiguration, resources api.ResourceCon
+	return []coreapi.Volume{}
+}
+
+func podVolume(config PodStepConfiguration) []coreapi.VolumeMount {
```
`s/podVolume
/secretVolumeMountFromConfig/`?

stevekuznetsov commented on pkg/steps/pod.go:
```diff
@@ -198,3 +171,82 @@ func PodStep(name string, config PodStepConfiguration, resources api.ResourceCon
+}
+
+func podSecret(config PodStepConfiguration) []coreapi.Volume {
+	if len(config.SecretName) > 0 {
```
Don't call unconditionally only to check if you should have been called at all. Above, do:

```go
pod := &coreapi.Pod{/* ... */}
if s.config.SecretName > 0 {
    pod.Spec.Containers[0].VolumeMounts = []coreapi.Volume{/* ... */}
}
```

stevekuznetsov commented on pkg/steps/pod.go:
```diff
@@ -198,3 +171,82 @@ func PodStep(name string, config PodStepConfiguration, resources api.ResourceCon
+func podVolume(config PodStepConfiguration) []coreapi.VolumeMount {
+	var secretPath string
+	if len(config.SecretName) > 0 {
+		if len(config.SecretPath) > 0 {
```
More idiomatic in go to do:

```go
secretPath := config.SecretPath
if secretPath == "" {
    secretPath = testSecretDefaultPath
}
```

stevekuznetsov commented on pkg/steps/pod_test.go:
```diff
@@ -182,3 +184,135 @@ func TestPodStepExecution(t *testing.T) {
+		expectedPod func(*v1.Pod)
+	}{
+		{
+			name: "no secret",
```
Give us cause and effect in test names please 

`"no secret name results in no mounted secrets"`

stevekuznetsov commented on pkg/steps/pod_test.go:
```diff
@@ -182,3 +184,135 @@ func TestPodStepExecution(t *testing.T) {
+				t.Errorf("test case %s error %v", tc.name, err)
+			}
+
+			if !reflect.DeepEqual(pod, podTemplate) {
```
Use `equality.Semantic.DeepEqual()` for k8s types

stevekuznetsov commented on pkg/steps/pod_test.go:
```diff
@@ -182,3 +184,135 @@ func TestPodStepExecution(t *testing.T) {
+
+			if !reflect.DeepEqual(pod, podTemplate) {
+				//if true {
+				t.Errorf("test %s failed. generated pod spec was not as expected", tc.name)
```
Use `diff.ObjectReflectDiff` for diffing

stevekuznetsov commented on pkg/steps/pod_test.go:
```diff
@@ -182,3 +184,135 @@ func TestPodStepExecution(t *testing.T) {
+	testCases := []struct {
+		name        string
+		podStep     func(*podStep)
+		expectedPod func(*v1.Pod)
```
This might be easier to read as a `func(*v1.Pod, *testing.T)` and just have this func validate the pod itself

stevekuznetsov replied:
Also technically you only validate the podspec, right?

mjudeikis replied:
We validate podSpec with PodTemplate which is mutated manually in the test scenario. So it validated function generated podSpec with our template. I think the test will become more complex if I stick with one function.

mjudeikis replied:
I still need to "construct reference point" so second function is needed. I could remove all `func expectedPodTemplate() *v1.Pod` but this means this test will become unextendable and very narrow. With current implementation it can be extended to test any podSpec mutation.

stevekuznetsov replied:
Instead of taking a template, mutating it and doing a DeepEqual, just validate the fields you need. Don't worry about this not being extensible, infinitely extensible unit tests aren't something we really need to be worrying about

stevekuznetsov commented on pkg/steps/pod_test.go:
```diff
@@ -182,3 +184,135 @@ func TestPodStepExecution(t *testing.T) {
+
+}
+
+func expectedPodTemplate() *v1.Pod {
```
If you only check the fields you care about your test will not need this, will not be convoluted with other behavior of the system and will be more concise

mjudeikis replied:
having a hard time to imagine how this could be squashed.

stevekuznetsov replied:
None of your tests care about the labels, or the annotations, or the command .. or ... or ... 

If you follow my suggestion and don't use `DeepEqual` but instead validate the specific fields you care about (feel free to `DeepEqual` on the  volume mounts, if you want) you don't need to present any extraneous fields and you don't need this func

droslean commented on pkg/steps/pod.go:
```diff
@@ -198,3 +170,79 @@ func PodStep(name string, config PodStepConfiguration, resources api.ResourceCon
+			Name: testSecretName,
+			VolumeSource: coreapi.VolumeSource{
+				Secret: &coreapi.SecretVolumeSource{
+					SecretName: config.SecretName,
```
pass only the `secretName` instead of the whole `PodStepConfiguration`

mjudeikis replied:
I prefer this to be more unified so both functions take the same input... But change as per suggestion.

droslean commented on pkg/steps/pod.go:
```diff
@@ -198,3 +170,79 @@ func PodStep(name string, config PodStepConfiguration, resources api.ResourceCon
+
+func secretVolumeMountFromConfig(config PodStepConfiguration) []coreapi.VolumeMount {
+	secretMountPath := config.SecretMountPath
+	if config.SecretMountPath == "" {
```
same here

stevekuznetsov commented:
What did you do to vendor? That diff doesn't look appropriate...

stevekuznetsov commented on pkg/api/types.go:
```diff
@@ -293,6 +297,13 @@ type TestStepConfiguration struct {
+// Secret describes a secret to be mounted inside a test
+// container.
+type Secret struct {
+	Name      string `json:"name"`
```
This is where people will go after the README for docs on the fields, so let's make sure they are documented here

mjudeikis commented:
> What did you do to vendor? That diff doesn't look appropriate...

add k8s validation, and glide install ... check the code for now, and I will try to drop it and redo...

stevekuznetsov commented on pkg/steps/pod.go:
```diff
@@ -198,3 +170,78 @@ func PodStep(name string, config PodStepConfiguration, resources api.ResourceCon
 	}
 }
+
+func (s *podStep) getPodObject() (*coreapi.Pod, error) {
```
This could use a more descriptive name than `getPodObject`

stevekuznetsov replied:
`generatePodForStep`? `generatePodForConfig`?

stevekuznetsov commented on pkg/steps/pod_test.go:
```diff
@@ -182,3 +184,133 @@ func TestPodStepExecution(t *testing.T) {
+			podStep: func(expectedPodStepTemplate *podStep) {
+				expectedPodStepTemplate.config.SecretName = ""
+			},
+			expectedPod: func(expectedPodTemplate *v1.Pod) {
```
So for instance ... right now to understand this test case I need to understand the template and then understand how the mutation is working. Instead, you can do a in-place validation with a `func(*coreapi.Pod, *testing.T)`:

```go
expectedPod: func(input *coreapi.Pod, t *testing.T) {
    if len(input.Spec.Volumes) != 0 {
        t.Errorf("declared volumes when none were expected: %v", diff.ObjectReflectDiff([]coreapi.Volume{}, input.Spec.Volumes)
    }
    if len(input.Spec.Containers[0].VolumeMounts) != 0 {
        t.Errorf("mounted volumes when none were expected: %v", diff.ObjectReflectDiff([]coreapi.VolumeMount{}, input.Spec.Containers[0].VolumeMounts)
    }
}
```


expected

mjudeikis replied:
I like your minimalistic style, but at this place, we want to know if what we set is per expectation. Not only if it is set. Refactored to middle ground what was before and what are you suggesting.

stevekuznetsov commented on pkg/steps/pod_test.go:
```diff
@@ -1,8 +1,10 @@
+	"path/filepath"
 	"testing"
 
+	coreapi "k8s.io/api/core/v1"
```
can you make sure you're always using `coreapi` and not `v1`?

mjudeikis commented:
Glide is so breaking vendoring... :/ cant add one dependency without updating other... and you point to master `openshift/api` and `client-go` :/ still working

stevekuznetsov commented on pkg/steps/pod_test.go:
```diff
@@ -182,3 +182,134 @@ func TestPodStepExecution(t *testing.T) {
+			podStep: func(expectedPodStepTemplate *podStep) {
+				expectedPodStepTemplate.config.SecretName = ""
+			},
+			expectedVolumeConfig: &v1.Pod{
```
These look good

mjudeikis commented:
final changes are done, and squashed.

stevekuznetsov commented:
@petr-muller can you take a final pass here?

droslean commented on pkg/steps/pod.go:
```diff
@@ -198,3 +182,65 @@ func PodStep(name string, config PodStepConfiguration, resources api.ResourceCon
+	return pod, nil
+}
+
+func secretVolumeFromConfig(secretName string) []coreapi.Volume {
```
nit: maybe the name `getVolumeFromSecret` or `getSecretVolume` ?

stevekuznetsov replied:
(lol that I suggested the other one)

droslean commented on pkg/steps/pod.go:
```diff
@@ -198,3 +182,65 @@ func PodStep(name string, config PodStepConfiguration, resources api.ResourceCon
+	}
+}
+
+func secretVolumeMountFromConfig(secretMountPath string) []coreapi.VolumeMount {
```
nit: likewise here, maybe `getSecretVolumeMount`?

droslean commented on pkg/steps/pod.go:
```diff
@@ -17,12 +17,17 @@ import (
 	Commands           string
 	ArtifactDir        string
 	ServiceAccountName string
+	SecretName         string
```
why not pass the whole `Secret` struct here?

mjudeikis replied:
Changed.

droslean commented on pkg/steps/pod.go:
```diff
@@ -198,3 +182,65 @@ func PodStep(name string, config PodStepConfiguration, resources api.ResourceCon
+		},
+	}
+
+	if s.config.SecretName != "" {
```
IMHO since we will always try to get both `volume` and `volumeMount`, I think it's better to have them in one function that returns both of values, and the nil check can be done by passing the whole `Secret` struct in the config.

stevekuznetsov replied:
Will we always have both? I thought we convenience default the mount path

mjudeikis replied:
I think it just overengineering. Normally we should always have both. But for this use-case. Future might change.

droslean replied:
No problem from my side. That was just a nit.

petr-muller commented:
I wasn't paying much attention to this, looked like @droslean is taking care of it. I feel like after ~90 comments it should be reviewed to death, don't block on me ;)

stevekuznetsov commented:
Ah sorry
@droslean okay, your call

droslean commented on pkg/steps/pod.go:
```diff
@@ -54,35 +58,12 @@ func (s *podStep) Run(ctx context.Context, dry bool) error {
+	if err != nil {
+		return err
 	}
+	pod.Spec.Containers[0].Image = image
```
nit: why not passing the `image` and `containerResources` in `generatePodForStep()`

droslean commented:
Please update also the https://github.com/openshift/ci-operator/blob/master/CONFIGURATION.md#ci-operator-configuration-reference

mjudeikis commented:
done

droslean commented on pkg/steps/pod.go:
```diff
@@ -198,3 +178,67 @@ func PodStep(name string, config PodStepConfiguration, resources api.ResourceCon
 	}
 }
+
+func (s *podStep) generatePodForStep(image string, containerResources coreapi.ResourceRequirements) (*coreapi.Pod, error) {
```
here you return `error` but it will always be `nil` since you are not checking for any errors inside this function.

droslean commented on pkg/steps/pod.go:
```diff
@@ -54,34 +58,9 @@ func (s *podStep) Run(ctx context.Context, dry bool) error {
-			},
-		},
+	pod, err := s.generatePodForStep(image, containerResources)
+	if err != nil {
```
this will always be nil

droslean commented:
/hold 
The pod will try to mount a secret that doesn't exist in the temporary namespace.  We have to update prowgen to add the `--secret-dir` arg and inject the secret in the prowjob's pod, or make the ci-operator to create the secret inside the temporary namespace.

I already discuss with @petr-muller , and we prefer the 2nd solution, as we don't want to involve `prowgen`

@stevekuznetsov @bbguimaraes @petr-muller  Any thoughts?

stevekuznetsov commented:
Why shouldn't we use the same secret copying mechanism as `--secret-dir`? It solves the same problem, right?

droslean commented:
`--secret-dir` requires the secrets to already be mounted in the pod (extra work for prowgen). But despite that, our case is simpler, since we want to mount a single secret. ci-operator can just get the secret from one namespace and create it to the temporary namespace.
example:
```yaml
secret:
  name: secret
  mount_path: /usr/secrets
  namespace: ci
```
Then ci-operator will get the secret from `ci` namespace and re-create it in the temporary namespace, then it will mount it in the specific test pod.

or
we merge that PR, and update the documentation that if anyone wants to use this feature he has to inject his secret in the prowjob and add the `--secret-dir` arg or make prowgen to do that.

I prefer the first sollution.  @stevekuznetsov thoughts?

stevekuznetsov commented:
Doing that would make it impossible to run `ci-operator` locally without access to read secrets in the `ci` namespace -- nobody has that. The `--secret-dir` impl is to make it so that the inputs for ci-operator can be provided literally.

/cc @smarterclayton

mjudeikis commented:
> Doing that would make it impossible to run `ci-operator` locally without access to read secrets in the `ci` namespace -- nobody has that. The `--secret-dir` impl is to make it so that the inputs for ci-operator can be provided literally.
> 
> /cc @smarterclayton

Agree. This is why I would say that adding this should be users responsibility (we already doing this). its one of those questions on how far we are willing to go :)

droslean commented:
Great. Let's also update the documentation in this PR.

smarterclayton commented:
In general the only secrets are the ones we manage, and profile is
supposed to be how we bundle them.  That keeps complexity low - single
input per build job.

We do need to think about how git secrets are provided, but that
wouldn’t be on the test, it would be in CI op and copied into the
namespace.

droslean commented:
/hold cancel

droslean commented:
/lgtm

droslean commented:
/approve
